docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+   particular, the dunder methods are described in `Version 3 of the NumPy Array
+   Interface`_. The exact order in which NumPy attempts to convert a foreign
+   object is described in the `interoperability with NumPy`_ document.
+
+If it is absolutely necessary to write functions where the input objects are not
+unconditionally known to be either PyTorch tensors or NumPy arrays, it is
+possible to ensure operator functionality by using NumPy functions explicitly as
+they are more forgiving than their PyTorch equivalents.
+
+.. csv-table::
+   :header: Operator, NumPy Function, Description
+
+   "``+``", "``np.add()``", "Addition"
+   "``-``", "``np.subtract()``", "Subtraction"
+   "``*``", "``np.multiply()``", "Multiplication"
+   "``/``", "``np.divide()``", "Division"
+
+Conversions
+-----------
+
+Only a small subset of data type (``dtype``) objects defined in NumPy have an
+equivalent in PyTorch, namely:
+
+.. csv-table:: $ indicates the sizes supported, e.g. ``uint8``
+   :header: ``np.dtype``, ``torch.dtype``, sizes
+
+    "``bool_``", "``bool``", "N/A"
+    "``uint$``", "``uint$``", ":math:`8`"
+    "``int$``", "``int$``", ":math:`8, 16, 32, 64`"
+    "``float$``", "``float$``", ":math:`16, 32, 64`"
+    "``complex$``", "``complex$``", ":math:`64, 128`"
+
+To ``numpy``
+^^^^^^^^^^^^
+
+The restrictions on a PyTorch tensor becoming a NumPy ``ndarray`` are:
+
+- It must be a dense tensor
+- It must be on the CPU
+- It must not have a computational graph
+- It must not have the conjugate bit set
+- It must not have the negative bit set
+- It must not be a subclass
+
+Essentially these can be expressed as:
+
+.. code-block:: python
+
+   # t is a torch.Tensor
+   assert(t.layout == torch.strided) # Dense
+   assert(t.is_cuda == False) # CPU
+   assert(t.requires_grad == False) # No autograd
+   assert(t.is_conj() == False) # Not conjugate
+   assert(t.is_neg() == False) # Not negative
+
+The :meth:`torch.numpy()` method  and the :doc:`np.asarray()
+<numpy:reference/generated/numpy.asarray>` function returns a **view** of the
+underlying tensor as a ``np.ndarray`` object.
+
+.. doctest::
+
+    >>> b_torch = torch.ones(3)
+    >>> b_torch.numpy()[2] = 32
+    >>> b_torch
+    tensor([ 1.,  1., 32.])
+    >>> a_np = np.array([1, 1, 32], dtype = np.float32)
+    >>> np.array_equal(b_torch.numpy(), a_np) # True
+    True
+    >>> c_tmp = np.asarray(b_torch, dtype = np.float32) # No copy if same dtype
+    >>> c_tmp
+    array([ 1.,  1., 32.], dtype=float32)
+    >>> c_tmp[2] = 1.
+    >>> b_torch
+    tensor([1., 1., 1.])
+
+.. note::
+
+   Since ``np.asarray()`` depends on the implementation of
+   ``torch.Tensor.__array__()`` which calls ``torch.numpy()``, the **leaf node**
+   requirement still needs to be satisfied by the user, i.e., ``requires_grad ==
+   False``
+
+From ``numpy``
+^^^^^^^^^^^^^^
+
+For a NumPy ``ndarray`` to be convertible to a PyTorch tensor:
+
+- It must have only native byte order
+- Must not be read only
```

Comment:
Don't we allow the conversion of non-writable arrays?

Reply:
From `tensor_numpy.cpp` in `csrc/utils/`, it **is** supported but throws a warning:

```
  TORCH_WARN_ONCE(
    "The given NumPy array is not writable, and PyTorch does "
    "not support non-writable tensors. This means writing to this tensor "
    "will result in undefined behavior. "
    "You may want to copy the array to protect its data or make it writable "
    "before converting it to a tensor. This type of warning will be "
    "suppressed for the rest of this program.");
}
```

I'll reword this to strongly suggest that the array should be writable.